forch/endpoint_handler.py

```diff
@@ -13,54 +13,100 @@
         self._logger = get_logger('endpproxy')
+        self._lock = threading.RLock()
+        self._mac_t1sw1_port = {}
+        self._freed_t1sw_ports = set()
+        self._next_t1sw_port = BASE_T1SW_PORT
+        self._structural_config_file = structural_config_file
         server_port = DEFAULT_SERVER_PORT
         address = f'{target_ip}:{server_port}'
-        self._logger.info('Proxy requests to %s, proto %s', address, PROTO_LOADED)
+        self._logger.info('Proxy requests to %s', address)
         channel = grpc.insecure_channel(address)
-        self._stub = server_grpc.SessionServerStub(channel)
+        self._stub = server_grpc.EndpointServerStub(channel)
         grpc.channel_ready_future(channel).result(timeout=CONNECT_TIMEOUT_SEC)
 
-    def process_endpoint(self, endpoint):
+    def _allocate_t1sw_port(self, mac):
+        with self._lock:
+            if mac in self._mac_t1sw1_port:
+                return self._mac_t1sw1_port[mac]
+            if self._freed_t1sw_ports:
+                t1sw_port = max(self._freed_t1sw_ports)
+                self._freed_t1sw_ports.remove(t1sw_port)
+            else:
+                t1sw_port = self._next_t1sw_port
+                self._next_t1sw_port += 1
+            self._mac_t1sw1_port[mac] = t1sw_port
+            with open(self._structural_config_file, 'r') as file:
+                structural_config = yaml.safe_load(file)
+                structural_config['dps'][T1_SW1]['interfaces'][t1sw_port] = {
+                    'description': 'TAP',
+                    'tagged_vlans': [171]
+                }
+            with open(self._structural_config_file, 'w') as file:
+                yaml.dump(structural_config, file)
+            return t1sw_port
+
+    def process_endpoint(self, endpoint, mac):
         """Handle an endpoint request"""
         self._logger.info('Process request for %s', endpoint.ip)
-        session_params = SessionParams()
-        session_params.endpoint.CopyFrom(endpoint)
-        for progress in self._stub.StartSession(session_params):
-            self._logger.info('Recevied progress %s', progress)
+        t1sw_port = self._allocate_t1sw_port(mac)
+        session_endpoint = Endpoint()
+        session_endpoint.ip = endpoint.ip
+        session_endpoint.port = endpoint.port
+        session_endpoint.vni = endpoint.vni
+        session_endpoint.tap_port = t1sw_port
+
+        self._stub.ConfigureInterface(session_endpoint)
         self._logger.info('Done with proxy request')
 
-
-class EndpointServicer(SessionServerServicer):
+    def free_endpoint(self, mac: str):
+        with self._lock:
+            self._logger.info('Process request to free endpoint for %s', mac)
+            freed_port = self._mac_t1sw1_port.pop(mac, None)
+            if not freed_port:
+                return
+            self._freed_t1sw_ports.add(freed_port)
+            with open(self._structural_config_file, 'r') as file:
```

Comment:
Can you extract the parts that deal with the structural config to a separate function?  It's mixing concepts here (endpoint management, with manipulating the file) so I think it'd be cleaner to factor out.

Personally I'd extract to one function (combining with above) that accepted a lambda for "do a think to this interface" that in one case would be setting it to something, and in the other case setting it to empty.  All the helper stuff around the file (safe_load, dump) would ideally only be in one place.